test/powershell/Modules/Microsoft.PowerShell.Core/ModuleConstraint.Tests.ps1

```diff
@@ -0,0 +1,1838 @@
+
+        $modSpec = New-ModuleSpecification -ModuleName $moduleName -ModuleVersion $ModuleVersion -MaximumVersion $MaximumVersion -RequiredVersion $RequiredVersion -Guid $Guid
+
+        $mod = Get-Module -FullyQualifiedName $modSpec
+
+        $mod | Should -Be $null
+    }
+
+    It "Does not load the module with FullyQualifiedName from absolute path when ModuleVersion=<ModuleVersion>, MaximumVersion=<MaximumVersion>, RequiredVersion=<RequiredVersion>, Guid=<Guid>" -TestCases $guidFailCases {
+        param($ModuleVersion, $MaximumVersion, $RequiredVersion, $Guid)
+
+        $modSpec = New-ModuleSpecification -ModuleName $modulePath -ModuleVersion $ModuleVersion -MaximumVersion $MaximumVersion -RequiredVersion $RequiredVersion -Guid $Guid
+
+        { Import-Module -FullyQualifiedName $modSpec -ErrorAction Stop } | Should -Throw -ErrorId 'Modules_ModuleWithVersionNotFound,Microsoft.PowerShell.Commands.ImportModuleCommand'
+    }
+
+    It "Does not load the module with FullyQualifiedName from the module path when ModuleVersion=<ModuleVersion>, MaximumVersion=<MaximumVersion>, RequiredVersion=<RequiredVersion>, Guid=<Guid>" -TestCases $guidFailCases {
+        param($ModuleVersion, $MaximumVersion, $RequiredVersion, $Guid)
+
+        $modSpec = New-ModuleSpecification -ModuleName $moduleName -ModuleVersion $ModuleVersion -MaximumVersion $MaximumVersion -RequiredVersion $RequiredVersion -Guid $Guid
+
+        { Import-Module -FullyQualifiedName $modSpec -ErrorAction Stop } | Should -Throw -ErrorId 'Modules_ModuleWithVersionNotFound,Microsoft.PowerShell.Commands.ImportModuleCommand'
+    }
+
+    It "Does not load the module with FullyQualifiedName from the manifest when ModuleVersion=<ModuleVersion>, MaximumVersion=<MaximumVersion>, RequiredVersion=<RequiredVersion>, Guid=<Guid>" -TestCases $guidFailCases {
+        param($ModuleVersion, $MaximumVersion, $RequiredVersion, $Guid)
+
+        $modSpec = New-ModuleSpecification -ModuleName $manifestPath -ModuleVersion $ModuleVersion -MaximumVersion $MaximumVersion -RequiredVersion $RequiredVersion -Guid $Guid
+
+        { Import-Module -FullyQualifiedName $modSpec -ErrorAction Stop } | Should -Throw -ErrorId 'Modules_ModuleWithVersionNotFound,Microsoft.PowerShell.Commands.ImportModuleCommand'
+    }
+
+    It "Does not load the module with version constraints from absolute path when ModuleVersion=<ModuleVersion>, MaximumVersion=<MaximumVersion>, RequiredVersion=<RequiredVersion>, Guid=<Guid>" -TestCases $failCases {
+        param($ModuleVersion, $MaximumVersion, $RequiredVersion, $Guid)
+
+        $sb = {
+            Invoke-ImportModule -Module $modulePath -MinimumVersion $ModuleVersion -MaximumVersion $MaximumVersion -RequiredVersion $RequiredVersion
+        }
+
+        if ($ModuleVersion -and $MaximumVersion -and ($ModuleVersion -ge $MaximumVersion))
+        {
+            $sb | Should -Throw -ErrorId 'ArgumentOutOfRange,Microsoft.PowerShell.Commands.ImportModuleCommand'
+            return
```

Comment:
why `return` instead of `else`?

Reply:
Given the choice I prefer `if`-`return`, because:
- The reader knows that they can stop reading when they see `return`
- [Less indentation](https://news.ycombinator.com/item?id=3414637) and less horizontal programming
- Closer to pattern match/guard style, which I find helps the "exhaust all possibilities" mentality
- The `return` works as a forcing function to not juggle state

I see many instances in the code base that I imagine initially started as innocent `if`-`else`s that could have been `if`-`return`s, but whose complexity has since increased due to that choice:
- https://github.com/PowerShell/PowerShell/blob/01634e4b2e71495933cc10c21744f2eaa7bed767/src/System.Management.Automation/engine/Modules/ImportModuleCommand.cs#L1698-L1783
- https://github.com/PowerShell/PowerShell/blob/01634e4b2e71495933cc10c21744f2eaa7bed767/src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs#L6351-L6506
- https://github.com/PowerShell/PowerShell/blob/01634e4b2e71495933cc10c21744f2eaa7bed767/src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs#L3646-L3715
- https://github.com/PowerShell/PowerShell/blob/c1c5344a8897262433141ecbc13bb06ac2c4bbef/src/System.Management.Automation/engine/Modules/ModuleSpecification.cs#L275-L325